poc/observe.py

```diff
@@ -0,0 +1,689 @@
+from functools import partial
+import itertools
+import logging
+import threading
+
+from traits import ctraits
+from traits.constants import ComparisonMode
+from traits.trait_base import Undefined, Uninitialized
+from traits.ctrait import CTrait
+from traits.trait_dict_object import TraitDictObject
+from traits.trait_set_object import TraitSetObject
+
+
+from poc.events import CTraitObserverEvent, ListObserverEvent
+from poc.interfaces import INotifiableObject
+from poc.trait_observer_notifier import (
+    TraitObserverNotifier,
+)
+from poc.listener_change_notifier import (
+    ListenerChangeNotifier,
+)
+
+# We need to identify objects which has this `_notifiers` methods
+# We could do the easy-to-ask-forgiveness-than-permission way.
+# Or we could do the leap-before-you-leap way.
+# Here is the LBYL way
+from traits.trait_list_object import TraitListObject
+INotifiableObject.register(TraitListObject)
+INotifiableObject.register(CTrait)
+INotifiableObject.register(ctraits.CHasTraits)
+
+logger = logging.getLogger()
+
+
+def dispatch_same(callback, args=(), kwargs=None):
+    if kwargs is None:
+        kwargs = {}
+    callback(*args, **kwargs)
+
+
+def dispatch_new_thread(callback, args=(), kwargs=None):
+    if kwargs is None:
+        kwargs = {}
+    threading.Thread(
+        target=callback, args=args, kwargs=kwargs).start()
+
+
+# These are user-facing notifier factories...
+DISPATCHERS = {
+    "extended": dispatch_same,
+    "same": dispatch_same,
+    "new": dispatch_new_thread,
+}
+
+
+def observe(object, callback, path, remove, dispatch):
```

Comment:
I think we want to make `remove` and `dispatch` keyword-only here (and perhaps some of the other arguments, too).

Could we also shuffle the signature a bit to place `object` and `path` together? IIUC, those two arguments comprise the "what" that we're observing